compiler/rustc_middle/src/mir/pretty.rs

```diff
@@ -82,34 +82,14 @@ pub fn dump_mir<'tcx, F>(
 ) where
     F: FnMut(PassWhere, &mut dyn Write) -> io::Result<()>,
 {
-    if !dump_enabled(tcx, pass_name, body.source.def_id()) {
-        return;
+    if let Some(filters) = &tcx.sess.opts.unstable_opts.dump_mir {
+        if pass_name_matches_dump_filters(tcx, filters, pass_name, body.source.def_id()) {
+            dump_mir(tcx, pass_num, pass_name, disambiguator, body, extra_data);
+        }
     }
-
-    dump_matched_mir_node(tcx, pass_num, pass_name, disambiguator, body, extra_data);
 }
 
-pub fn dump_enabled<'tcx>(tcx: TyCtxt<'tcx>, pass_name: &str, def_id: DefId) -> bool {
-    let Some(ref filters) = tcx.sess.opts.unstable_opts.dump_mir else {
-        return false;
-    };
```

Comment:
is it actually perf relevant to do this at every call site instead of inside the function?

Reply:
It is, as the perf CI results show.

There is a two-step check when deciding whether to dump MIR.
- Is `-Zdump-mir` specified?
- If so, does the MIR pass name match the `-Zdump-mir` value?

If both those conditions pass, then the MIR is dumped, which involves printing `pass_name` and `pass_num`.

For most of the `dump_mir` callsites `pass_name` and `pass_num` are simple scalars, which poses no problem.
But at at two of the `dump_mir` call sites (in ` dump_mir_for_phase_change` and `dump_mir_for_pass`) `pass_name` and `pass_num` are constructed in a way that is moderately expensive, and this happens before the two-step check.

The PR changes things so that, at two of these call sites, `pass_name` is only constructed if the first part of the two-step check succeeds, and `pass_num` is only constructed if the second part of the two-step check succeeds.

Reply:
I tried building an alternative design (https://github.com/rust-lang/rust/pull/105121) that afaict should end up with similar perf improvements.